### ISSUE TITLE
Chef Server Analytics Queue Depth Should Be Capped When No Consumer Present

### ISSUE THREAD

**Issue by sean-horn** [chef/chef-server#103]
As a customer, I want the chef server to be as resilient as possible to unexpected conditions.

An unexpected, but common condition that occurs is that the Analytics server goes offline and stops consuming messages from the "/analytics/alaska" and potentially other queues. When this happens, the messages build up without limits and eventually fill all available disk space and cause a failure of the chef server system, by swapping, rabbitmq paging out to disk, and other general expected unpleasantness.

To remedy this situation, whatever chef server component writes to these queues should stop writing if it detects no consumption and that the queue is growing.

Zendesk 3070 and https://github.com/chef/omnibus-analytics/issues/179

**Comment by marcparadise** [comment]
A better solution here may just be to set the rabbitmq max queue length for the relevant queues:  

http://www.rabbitmq.com/maxlength.html

**Comment by sean-horn** [comment]
Cool, sounds like a good addition for all queues. If the expander stops working for any reason, the incoming node.save and other objects can and have overrun chef server systems in the past. So, the issue is not limited to just Analytics.

**Comment by sean-horn** [comment]
This is currently implemented at a few customer sites

    rabbitmqctl set_policy -p /analytics max_length '(alaska|notifier.notifications|notifier_config)' '{"max-length":1000}' --apply-to queues

**Comment by metadave** [comment]
The problem with setting a max depth for the analytics queue is that we'll miss messages, which won't work for people that are using Analytics for compliance etc. I'm implementing a `opscode-analytics-ctl` plugin that will remove the queue and/or /analytics vhost and allow for reconfiguration if desired.

**Comment by stevendanna** [comment]
@metadave I may not be understanding all the details here.  How will these -ctl commands help us in the face of a analytics server that temporarily gone offline?  

What is the customer expected behavior?  If the customer expect that we will never lose analytics data even in the face of a network partition or temporary analytics outage, I think we need *something* that informs the chef-server "no more writes until analytics catches up."

**Comment by metadave** [comment]
uh, backpressure can be tricky.

The `opscode-analytics-ctl` commands are for users that want to *disable* analytics. For temporary outages, a max-depth is probably fine for users that aren't using compliance features. We'll have to give compliance some thought.

**Comment by sean-horn** [comment]
Zendesk 3070 , 3385, and 3297, all Sev1/2 have resulted from unlimited access to queues.

**Comment by metadave** [comment]
Cool, thanks for patching up customers on the Sev1/2's. We're committed to cleaning up scalability and reliability quite a bit this quarter.

**Comment by sean-horn** [comment]
@metadave jhs had the same thought. Compliance customers could be considered to be in a class of their own. It would seem everyone else's chef server would benefit from this addition. Then, we just allow opscode-omnibus changes when full durability is prescribed for the Compliance config case.

**Comment by jeremiahsnapp** [comment]
@metadave I spent a lot of time replicating this problem after one of the more recent severe outages and I completely agree with what @sean-horn is saying.

All rabbitmq queues need to be capped by default in order to provide the best experience with least surprise. This cap can still be reasonably large enough to avoid losing messages in most situations while still preventing worst case situations like filling the disk with persistent messages.

Capping the queues by default doesn't necessarily mean messages will be lost. Messages will only get lost if that cap is reached and if that happens then something else is wrong in the first place and needs to be addressed.

The Compliance/Audit use case should be treated as the exception and a tool should be created to disable the cap or adjust its size and Compliance/Audit related documentation should clearly point to the need for using that tool. The documentation should also make it clear the dangerous potential that awaits the uncapped queue. They potentially have a lot more to lose than messages with an uncapped queue.

In my opinion, the tool shouldn't even remove the cap. Even in the Compliance/Audit use case the cap could (should) still remain to safe guard against filling the disk or otherwise cause problems. Analytics consumers shouldn't be down for such a long time that the cap gets hit if the user really cares about the data. If consumers need to be down then the user should appropriately account for this by reducing the amount of messages created (e.g. reducing number of chef-client runs, etc) to avoid hitting the cap.

Without capped queues you run the risk of either filling disks and breaking the system in much worse ways or running into rabbitmq memory problems or eventually restarting rabbitmq only to find out it takes an unreasonable amount of time to start because it is reindexing hundreds of GBs of messages. All of these situations require unexpected downtime and some even raise the potential for corrupting data.

We have seen all of these situations in real world installs. Once these situations occur you simply can not manage rabbitmq queues because rabbitmq isn't able to fully start and bind to the management ports. In every instance I know of we had to resort to deleting the persistent messages from disk manually in order to restore a functional Chef Server. Obviously this resulted in lost messages anyway and a lot worse experience.
